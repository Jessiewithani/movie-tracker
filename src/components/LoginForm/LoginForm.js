import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';
import './LoginForm.scss';
import { logIn, setCurrentUser, saveFavorites } from '../../actions';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { loginUser, getFavorites } from '../../apiCalls/apiCalls';
import { favorites } from '../../reducers/favorites';

export class LoginForm extends Component {
  constructor() {
    super()
    this.state = {
      loginEmail: '',
      loginPassword: '',
      status: null,
      loginPasswordError: ''
    }
  }

  handleChange = (e) => {
    this.setState({ [e.target.name]: e.target.value });
  }

  clearLoginInputs = () => {
    this.setState({ loginEmail: '', loginPassword: '' })
  }

  loginUser = async e => {
    e.preventDefault();
    const { logIn, setCurrentUser, saveFavorites } = this.props;
    const { loginEmail, loginPassword } = this.state;
    const user = { email: loginEmail, password: loginPassword };
   
    try {
      let currentUser = await loginUser(user);
      let favorites = await getFavorites(currentUser.id);
      setCurrentUser(currentUser);
      saveFavorites(favorites);
      logIn();
      this.setState({status: 200, loginPasswordError: ""})
    } catch (error) {
      console.log(error.message)
      this.setState({status: 401, loginPasswordError: "* the password does not match! *"})
    }
    
    this.clearLoginInputs();
  }

  render() {
    if (this.state.status === 200) {
      return <Redirect to='/' />
    } 

    return (
      <div>
        <section className="login">
          <form>
            <input className="login-input" type="email" placeholder="insert login e-mail" name="loginEmail" value={this.state.loginEmail} onChange={this.handleChange} />
            <input className="login-input" type="password" placeholder="insert login password" name="loginPassword" value={this.state.loginPassword} onChange={this.handleChange} />
            <h4 style={{ color: "red" }}>{this.state.loginPasswordError}</h4>
            <button className="login-button a-login-button" onClick={(e) => this.loginUser(e)}>LOG IN</button>
          </form>
        </section>
      </div>
    )
  }
}

const mapStateToProps = ({ user }) => ({
  user
})

const mapDispatchToProps = dispatch => (
  bindActionCreators({
    logIn,
    setCurrentUser,
    saveFavorites
  }, dispatch)
)

export default connect(mapStateToProps, mapDispatchToProps)(LoginForm);