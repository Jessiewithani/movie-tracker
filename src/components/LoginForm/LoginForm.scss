.login {
    background: url('https://bbbslubbock.org/wp-content/uploads/2018/10/grey-background-for-site-for-website-1432015867.jpg');
    justify-content: center;
    align-items: center;
    display: flex;
    height: 100vh;
}

.login-input {
    font-size: 2.5em;
    font-family: 'Nanum Gothic', sans-serif;
    color: black;
    padding-left: 15px;
    border: none;
    border-bottom: solid white 5px;
    background-color: rgba(0, 0, 0, 0);
}

.login-button {
    font-size: 2em;
    font-family: 'Josefin Slab', serif;
    color: lightgrey;
    font-weight: bold;
    height: 2.5em;
    width: 700px;
    background-color: rgba(0, 0, 0, 0.5)
}

a.a-login-button:hover {
    background-color: white;
}