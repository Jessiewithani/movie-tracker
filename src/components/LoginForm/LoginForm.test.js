import { LoginForm, mapStateToProps } from './LoginForm';
import React from 'react';
import { shallow, mount } from 'enzyme';
import { MemoryRouter } from 'react-router';
import { Redirect } from 'react-router-dom'



describe('LoginForm', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <LoginForm 
      />)
  })
  
  it('should match the snapshot with all data passed in', () => {

    expect(wrapper).toMatchSnapshot()
  });
  it('should call loginUser when button is clicked', () => {

    wrapper.instance().loginUser = jest.fn()
    const mockEvent = { preventDefault: jest.fn()}

    wrapper.instance().forceUpdate();

    wrapper.find('button').simulate('click', mockEvent)

    expect(wrapper.instance().loginUser).toHaveBeenCalled()

  });
  it('should update state when handleChange is invoked', () => {
    const mockEvent = {target: {name: 'name', value: 'Victor'}}
    const expected = 'Victor'

    wrapper.instance().handleChange(mockEvent);
    expect(wrapper.state('name')).toEqual(expected)
  });
  it('should redirect to home if the user is successfully logged in', () => {
    let wrapper = mount(
      <MemoryRouter initialEntries={['/login']}>
        <Redirect to={'/'}/>
      </MemoryRouter>
    )

    expect(wrapper.find(Redirect)).toHaveLength(1)
  })
})

describe('mapStateToProps in LoginForm', () => {
  it('should return an object with user info', () => {
    const mockState = {
      user: {
        name: 'Eric',
        email: 'Eric@gmail.com',
        password: 'Eric234'
      },
      saveUser: 'SAVE_USER'
    }

    const expected = {
      user: {
        name: 'Eric',
        email: 'Eric@gmail.com',
        password: 'Eric234'
      }
    }

    const mappedProps = mapStateToProps(mockState);

    expect(mappedProps).toEqual(expected)
  })
})