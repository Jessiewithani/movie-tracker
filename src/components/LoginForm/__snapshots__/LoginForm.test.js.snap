// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should match the snapshot with all data passed in 1`] = `
<div>
  <section
    className="login"
  >
    <form>
      <input
        className="login-input"
        name="loginEmail"
        onChange={[Function]}
        placeholder="insert login e-mail"
        type="email"
        value=""
      />
      <input
        className="login-input"
        name="loginPassword"
        onChange={[Function]}
        placeholder="insert login password"
        type="password"
      />
      <h4
        style={
          Object {
            "color": "red",
          }
        }
      />
      <button
        className="login-button a-login-button"
        onClick={[Function]}
      >
        LOG IN
      </button>
    </form>
  </section>
</div>
`;
